providers/implementations/encode_decode/encode_key2any.c

```diff
@@ -777,127 +803,140 @@ static int key2any_set_ctx_params(void *vctx, const OSSL_PARAM params[])
-    impl##_##kind##2##output##_encode_o(void *vctx, const void *key,    \
-                                        OSSL_CORE_BIO *cout,            \
-                                        OSSL_PASSPHRASE_CALLBACK *cb,   \
-                                        void *cbarg)                    \
+    impl##2##output##_encode(void *ctx, OSSL_CORE_BIO *cout,            \
+                             const void *key,                           \
+                             const OSSL_PARAM key_abstract[],           \
+                             int selection,                             \
+                             OSSL_PASSPHRASE_CALLBACK *cb, void *cbarg) \
     {                                                                   \
-        int selection = type##_##kind##_selection;                      \
-                                                                        \
-        if (!ossl_assert(selection != 0)                                \
-            || !ossl_assert((selection & ~ALLOWED_SELECTORS) == 0)) {   \
-            ERR_raise(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR);              \
+        /* We don't deal with abstract objects */                       \
+        if (key_abstract != NULL) {                                     \
+            ERR_raise(ERR_LIB_PROV, ERR_R_PASSED_INVALID_ARGUMENT);     \
             return 0;                                                   \
         }                                                               \
         if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)         \
-            return key2any_encode(vctx, key, impl##_type_to_evp(key),   \
-                                  cout, key_to_##output##_pkcs8_bio,    \
+            return key2any_encode(ctx, cout, key, impl##_evp_type,      \
+                                  type##_check_key_type,                \
+                                  key_to_##output##_pkcs8_bio,          \
                                   cb, cbarg,                            \
                                   prepare_##type##_params,              \
                                   type##_priv_to_der);                  \
         if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)          \
-            return key2any_encode(vctx, key, impl##_type_to_evp(key),   \
-                                  cout, key_to_##output##_pubkey_bio,   \
+            return key2any_encode(ctx, cout, key, impl##_evp_type,      \
+                                  type##_check_key_type,                \
+                                  key_to_##output##_pubkey_bio,         \
                                   cb, cbarg,                            \
                                   prepare_##type##_params,              \
                                   type##_pub_to_der);                   \
-        return key2any_encode_params(vctx, key, cout,                   \
+        return key2any_encode_params(ctx, cout, key, impl##_evp_type,   \
```

Comment:
No selection check here??

Reply:
Didn't we have this discussion already?

Anyway, I added a check, and a return 0 at the end